feat(posthog): Add Telemetry client Posthog

javirln opened the pull request:
This PR creates a new implementation of a Tracker client based on Posthog.

Users can deactivate telemetry by setting the env variable `DO_NOT_TRACK=1`, it follows the convention at: https://consoledonottrack.com/

Information being tracked:
- System information (OS, ARCH)
- Command ran
- CI Runner information if detected
- CLI version
- Hashed Control Plane URL
- Anonymous device ID
- User and Organization if logged in

Refs: #780

migmartri commented on app/cli/cmd/root.go:
```diff
@@ -83,6 +93,38 @@ func NewRootCmd(l zerolog.Logger) *cobra.Command {
+				logger.Debug().Msg("Telemetry is disabled")
+			} else {
+				var err error
+				telemetryClient, err = posthog.NewClient(posthogAPIKey, posthogEndpoint)
```
so you do not want to crash on purpose correct?

javirln replied:
Yes, I don't want to crash if the telemetry fails, I think that could disturb users and automations

migmartri commented on app/cli/cmd/root.go:
```diff
@@ -83,6 +93,38 @@ func NewRootCmd(l zerolog.Logger) *cobra.Command {
 		PersistentPostRunE: func(cmd *cobra.Command, args []string) error {
+			// Do not track telemetry if the env variable is set
+			var telemetryClient telemetry.Client
+			if os.Getenv(doNotTrackEnv) == "1" || os.Getenv(doNotTrackEnv) == "true" {
```
it would be nice if the detection could be extracted as a function `isTelemetryDisabled() bool`

migmartri commented on app/cli/cmd/root.go:
```diff
@@ -83,6 +93,38 @@ func NewRootCmd(l zerolog.Logger) *cobra.Command {
+				currentCmd = currentCmd.Parent()
+			}
+
+			cmdLine := strings.Join(cmdHierarchy, " ")
```
and the command line generation also to be extracted to a function too so it could potentially be tested.

migmartri commented on app/cli/cmd/root.go:
```diff
@@ -83,6 +93,38 @@ func NewRootCmd(l zerolog.Logger) *cobra.Command {
+			cmdLine := strings.Join(cmdHierarchy, " ")
+
+			err := cmdTracker.Track(cmd.Context(), cmdLine, telemetry.Tags{
+				"cli_version": Version,
```
it would be great if we could also capture the conttrolplaneURL but hash it anonymously. We have the controlplane URI in viper, I guess we should be able to find it.

migmartri replied:
and also the command status. Isn't it a way to detect if the command returned an error? We capture errors somewhere else globally to do some manual handling, i.e show you that you need to re-login

javirln replied:
I have added both of them computing the hash of the control plane url

migmartri commented on app/cli/internal/telemetry/posthog/posthog.go:
```diff
@@ -53,14 +53,20 @@ type Tracker struct {
 
 // TrackEvent sends an event to the PostHog server.
 func (p *Tracker) TrackEvent(_ context.Context, eventName string, id string, tags telemetry.Tags) error {
+	defer p.client.Close()
```
why is this needed? To send the event once the function finishes?

javirln replied:
Yes and to ensure the client is properly closed.

migmartri commented on app/cli/cmd/root.go:
```diff
@@ -47,6 +50,13 @@ const (
 	appName                 = "chainloop"
 	//nolint:gosec
 	tokenEnvVarName = "CHAINLOOP_TOKEN"
+	doNotTrackEnv   = "DO_NOT_TRACK"
```
add a comment with a link to https://consoledonottrack.com/

migmartri commented on app/cli/cmd/root.go:
```diff
@@ -83,6 +93,38 @@ func NewRootCmd(l zerolog.Logger) *cobra.Command {
+			if os.Getenv(doNotTrackEnv) == "1" || os.Getenv(doNotTrackEnv) == "true" {
+				logger.Debug().Msg("Telemetry is disabled")
+			} else {
+				var err error
```
I'd add a debug statement saying "Telemetry enabled, to disable it use DO_NOT_TRACK=1"

migmartri commented on app/cli/cmd/root.go:
```diff
@@ -83,6 +93,38 @@ func NewRootCmd(l zerolog.Logger) *cobra.Command {
+			// Do not track telemetry if the env variable is set
+			var telemetryClient telemetry.Client
+			if os.Getenv(doNotTrackEnv) == "1" || os.Getenv(doNotTrackEnv) == "true" {
+				logger.Debug().Msg("Telemetry is disabled")
```
we might not need this debug statement but instead put one when it's enabled so they know how to disable it. It will also remove the if/else

migmartri reviewed (commented):
Cool!

javirln reviewed (commented):
Refactored the code to be able to check the exit code as it was not trivial since cobra hooks are not executed when a command fails.

In any case I think it's pretty neat, we have everything we wanted.

jiparis commented on app/cli/main.go:
```diff
@@ -16,25 +16,75 @@
-		os.Exit(exitCode)
+	}
+
+	runTelemetry(logger, executedCmd, exitCode)
```
I see that only succeeded commands will be tracked.
Do you thins this could be moved to the `PreRunE` (or PostRun) method in the rootCmd command instead?

javirln replied:
No no, it will track both successful and unsuccessful commands that's why it's moved into the main package. The problem with the `PostRun(E)` is that they don't run if there has been an error. Moving the logic to the main caller, `rootCmd.Execute` give us the chance to hook it up to know when there has been an error and track it as well.

Additionally to know what's the actual command that has been run we moved it to use `rootCmd.ExecuteC` which returns the executed command therefore in this way we have the whole picture :)

migmartri replied:
but to be clear, there is a bug in that code that doesn't do that right? That was a catch from @jiparis.

jiparis replied:
Sorry, my bad, I didn't notice the removal of `os.Exit(exitCode)`. Ignore my comment.

javirln commented:
Secrets on the repository updated already to the proper project.

migmartri commented on .goreleaser.yml:
```diff
@@ -28,6 +28,8 @@ builds:
     ldflags:
       - "{{ .Env.COMMON_LDFLAGS }}"
       - -X github.com/chainloop-dev/chainloop/app/cli/cmd.Version={{ .Version }}
+      - -X main.posthogAPIKey={{ .Env.POSTHOG_API_KEY }}
```
Let's make sure we have a page documenting this change that goes up in the same Chainloop release :)

javirln replied:
Yes, I still have pending add a documentation page.

migmartri commented on app/cli/cmd/root.go:
```diff
@@ -269,12 +269,14 @@ func parseToken(token string) (*ParsedToken, error) {
 		switch aud[0].(string) {
 		case apiTokenAudience:
-			return &ParsedToken{Type: "api-token"}, nil
+			tokenID := claims["jti"].(string)
```
you might want to protect this casting a little bit
```
tokenID, ok := claims ...
if !ok {
return nil, nil
}
```

migmartri commented on app/cli/cmd/root_test.go:
```diff
@@ -31,6 +31,7 @@ func TestParseToken(t *testing.T) {
 			name:  "robot account",
 			token: "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJvcmdfaWQiOiI5M2QwMjI3NS04NTNjLTRhZDYtOWQ2MC04ZjU2MmIxMjNmZDIiLCJ3b3JrZmxvd19pZCI6IjM1ZTZkOGIwLWE0OGYtNDFmYS05YmU3LWQ1OTM5YjJkZGUyNiIsImlzcyI6ImNwLmNoYWlubG9vcCIsImF1ZCI6WyJhdHRlc3RhdGlvbnMuY2hhaW5sb29wIl0sImp0aSI6IjQ4YWVhMWNiLTk5MGUtNDM2OS1hOTFhLTczNTIzNzk1NjhiNSJ9.aYPAPK-AauJpxRGEapV2ejwxhyhmFej6S79Ni-xJ4Q0",
 			want: ParsedToken{
+				ID:   "48aea1cb-990e-4369-a91a-7352379568b5",
```
thanks for adding tests.

migmartri commented on app/cli/cmd/root.go:
```diff
@@ -269,12 +269,14 @@ func parseToken(token string) (*ParsedToken, error) {
 		case robotAccountAudience:
-			return &ParsedToken{Type: "robot-account"}, nil
+			tokenID := claims["jti"].(string)
+			return &ParsedToken{Type: "robot-account", ID: tokenID}, nil
```
btw, organization has been added and will be available in both robot-account and api-token once #795 is merged :)

javirln replied:
Awesome

migmartri commented on app/cli/main.go:
```diff
@@ -16,25 +16,82 @@
-	if err := rootCmd.Execute(); err != nil {
-		msg, exitCode := errorInfo(err, logger)
+	// Set the default exit code to 0, in case of an error it will be updated
+	exitCode := 0
```
nitpick. Leveraging default values would have helped in this case too. i.e `var exitCode int`

javirln replied:
Sure.

migmartri commented on app/cli/main.go:
```diff
@@ -16,25 +16,82 @@
+
+	runTelemetry(logger, executedCmd, exitCode)
+
+	os.Exit(exitCode)
```
do we want to `os.Exit` even if we have 0 as returned code? I know it's the last statement of the program but I am not sure if there is an additional consequence of doing this instead of letting the program finish as before

javirln replied:
I know but the other option I think it's weird:
```golang
if !isTelemetryDisabled() {
  recordCommand(logger, executedCmd, exitCode)
}

if exitCode != 0 {
  os.Exit(exitCode)
}
```

migmartri replied:
yeah, I agree, my question was more out of curiosity, to know if there was a difference between letting the program finish or calling exit(0), it might not be.

javirln replied:
Ahh sorry! To be honest I'm not sure, I guess it won't have any different :/

migmartri commented on app/cli/main.go:
```diff
@@ -16,25 +16,82 @@
+}
+
+// runTelemetry sends the telemetry data to the configured endpoint
+func runTelemetry(logger zerolog.Logger, executedCmd *cobra.Command, exitCode int) {
```
I'd call this method, `recordCommand` to be more specific of the scope of this telemetry

migmartri commented on app/cli/main.go:
```diff
@@ -16,25 +16,82 @@
+func runTelemetry(logger zerolog.Logger, executedCmd *cobra.Command, exitCode int) {
+	var telemetryClient telemetry.Client
+
+	if !isTelemetryDisabled() {
```
tbh, at this point is probably better that we wrap this `runTelemetry` method within `!isTelemetryDisabled`, that way we do not run any useless compute like generating all those tags and so on if we know it's not going to happen.

In other words, I'd do something like

if !isTelemetryDisabled() { runTelemetry(...) }

migmartri commented on app/cli/main.go:
```diff
@@ -16,25 +16,82 @@
+		var err error
+		telemetryClient, err = posthog.NewClient(posthogAPIKey, posthogEndpoint)
+		if err != nil {
+			logger.Debug().Err(err).Msgf("creating telemetry client: %v", err)
```
why don't we return here?

javirln replied:
We actually could, I'll change it.

migmartri commented on app/cli/main.go:
```diff
@@ -16,25 +16,82 @@
+
+	err := cmdTracker.Track(executedCmd.Context(), extractCmdLineFromCommand(executedCmd), tags)
+	if err != nil {
+		logger.Debug().Err(err).Msgf("tracking command: %v", err)
```
same here, we could return technically

migmartri commented on app/cli/internal/telemetry/telemetry.go:
```diff
@@ -49,28 +50,56 @@ func NewCommandTracker(client Client) *CommandTracker {
+}
+
+// Determines the user ID using available information or generates a random one.
+func determineUserID(tags map[string]string) string {
```
it would be great to add some tests for this method to help explain what it does.

migmartri commented on app/cli/internal/telemetry/telemetry.go:
```diff
@@ -49,28 +50,56 @@ func NewCommandTracker(client Client) *CommandTracker {
+	machineID, _ := machineid.ProtectedID("chainloop")
+	tags["machine_id"] = machineID
+
+	// Check if user ID is provided in tags.
```
add a comment saying that this is not the case on the unauthenticated scenario

migmartri commented on app/cli/main.go:
```diff
@@ -16,25 +16,82 @@
+		"source":      "cli",
+	}
+
+	if parsedToken := executedCmd.Context().Value(cmd.AuthenticationToken{}); parsedToken != nil {
```
add a comment saying that this context information might not be here if the user is not authenticated, and that's ok, expected and handled by the underlying library.

migmartri reviewed (approved):
Added some minor comments :)

migmartri commented on app/cli/main.go:
```diff
@@ -16,25 +16,82 @@
+		"cli_version": cmd.Version,
+		"exit_code":   fmt.Sprint(exitCode),
+		"cp_url_hash": hashControlPlaneURL(),
+		"source":      "cli",
```
I'd rename this to `chainloop_source` so we can differentiate it from other common analytic property. What do you think?

javirln replied:
Sure

migmartri commented on app/cli/main.go:
```diff
@@ -16,25 +16,86 @@
+func recordCommand(logger zerolog.Logger, executedCmd *cobra.Command, exitCode int) {
+	telemetryClient, err := posthog.NewClient(posthogAPIKey, posthogEndpoint)
+	if err != nil {
+		logger.Debug().Err(err).Msgf("creating telemetry client: %v", err)
```
this debug entry seems to be shown even if debug level is not enabled. 

```
./bin/chainloop wf ls
ERR authentication required, please run "chainloop auth login"
DBG creating telemetry client: invalid configuration, API Key and endpoint URL are required error="invalid configuration, API Key and endpoint URL are required"
```

not sure why, or if this is the expected behavior.

migmartri replied:
ok, the problem is that this logger is not correctly initialized the one in root is

```
func initLogger(logger zerolog.Logger) (zerolog.Logger, error) {
	lvl := zerolog.InfoLevel
	if flagDebug {
		lvl = zerolog.DebugLevel
	}

	return logger.Level(lvl), nil
}
```

This means that the logger I believe is in debug mode out of the box, not great I think. Not sure how to solve it though. At least we should make sure that these errors that are not meant to scare the user are not shown.

migmartri replied:
I can see that a way to do this would be to newRootCommand to return also the initialized logger, to main can use it accordingly

migmartri replied:
not trivial, we could leave it for now but create a task to propagate the level to main, wdyt?

javirln replied:
Yeah...I was giving a try this morning precisely with what you mention of returning the logger but it seems it does not take into account the previously set level...
I'll create a task for it.

migmartri replied:
well,not really, do you also see this message on every command?

```
./bin/chainloop wf ls                                                  
DBG Telemetry enabled, to disable it use DO_NOT_TRACK=1
```

javirln replied:
Yes in all commands, it does not matter the value of the debug flag it will always be shown.

migmartri commented on app/cli/main.go:
```diff
@@ -16,25 +16,86 @@
 package main
 
 import (
+	"crypto/sha256"
```
could you bump the year on the header?

javirln replied:
Sure

migmartri reviewed (changes requested):
Do not show debug messages by default, especially the one about telemetry enabled

javirln commented:
> Do not show debug messages by default, especially the one about telemetry enabled

I've moved the message to the `root` command. I was not able to set the proper flag from outside not making the logger to be consistent with the logging level. I even tried the global setting from zerlog without luck.

migmartri reviewed (commented):
Thanks,

I tried to figure out the `flag` issue and boy it's quite confusing, ..., I kinda managed to fix it but I do not like my solution. In am ok going with your solution for now but we should add a task to fix it. 

Re telemetry performance, I'll take a quick look and see if I find anything, once done I think I'll be ready to approve this and move forward. Thanks!

migmartri commented on app/cli/internal/telemetry/telemetry.go:
```diff
@@ -49,28 +50,57 @@ func NewCommandTracker(client Client) *CommandTracker {
+}
+
+// DetermineUserID Determines the user ID using available information or generates a random one.
+func DetermineUserID(tags Tags) string {
```
this function probably doesn't need to be exposed. I am assuming it is because the test package uses `_test`. In that case I'd ask the question if we need the `test` file to use a different package or just treat it as unit tests in the `telemetry` package.

What do you think?

migmartri commented on app/cli/main.go:
```diff
@@ -16,25 +16,86 @@
+	if contextToken != nil {
+		parsedToken, ok := contextToken.(*cmd.ParsedToken)
+		if ok {
+			tags["token_type"] = parsedToken.Type
```
I get a panic when using locally a CHAINLOOP_API_TOKEN during an attestation

```
./bin/chainloop att init --replace
ERR rpc error: code = InvalidArgument desc = when using an API Token, workflow name is required as parameter
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x201ea00]

goroutine 1 [running]:
main.recordCommand(0xc000f47508, 0x1, 0x18fffa5e)
        /home/migmartri/work/chainloop/chainloop/app/cli/main.go:91 +0x3a0
main.main()
        /home/migmartri/work/chainloop/chainloop/app/cli/main.go:64 +0x2c8
```

migmartri replied:
Ok, the problem was that my token didn't have `org_id` since the current live backend doesn't have it yet. 

but I think this is surfacing another problem, that the `parseToken` method can return 3 values, `nil`, empty token `&ParsedToken{}` or a full-fledged token. 

We should probably only return `nil` or a token and handle it accordingly.

migmartri commented on app/cli/cmd/root.go:
```diff
@@ -269,12 +283,31 @@ func parseToken(token string) (*ParsedToken, error) {
 		switch aud[0].(string) {
 		case apiTokenAudience:
-			return &ParsedToken{Type: "api-token"}, nil
+			tokenID, ok := claims["jti"].(string)
```
related to another of my comments, I think we should improve this method, it relies too much in casting, causing for example to return `nil` if the token doesn't have a property.

I think this could be improved by 

a - leveraging `claims.VerifyAudience(aud, true) bool`
b - once found, we could map to their actual claims so they are no longer a map of interface and can be accessed properly. 

I think this way the function will be less brittle, what do you think?

migmartri commented on app/cli/cmd/root_test.go:
```diff
@@ -44,8 +45,9 @@ func TestParseToken(t *testing.T) {
 		},
 		{
 			name:  "api token",
```
you should probably try a token that's missing some properties. That's a case that could happen in the future and will happen today (old API tokens without org)

javirln commented:
Refactored all code with:
- Moved all logic from `main` to `root`.
- Use a separated go routine to send the telemetry on the `persistentPreRunE`. Use a `sync.WaitGroup` to wait for it in case it's slow.
- Refactored pase of the jwt token in a generic way, otherwise we would have to cast the token to each type of claim coupling backend and CLI due to types being on the backend.
- Added more tests with empty fields and random tokens.

migmartri commented on app/cli/main.go:
```diff
@@ -31,9 +31,12 @@ func main() {
 
+	// Run the command
 	if err := rootCmd.Execute(); err != nil {
+		var msg string
```
we can probably revert all the changes in this file. For example, having the msg var  doesn't seem to make sense anymore.

migmartri commented on app/cli/cmd/root.go:
```diff
@@ -53,12 +57,16 @@ const (
+var telemetryWg sync.WaitGroup
+
+type parsedToken struct {
+	ID    string
```
just noticed, these properties do not need to be uppercase either

migmartri commented on app/cli/cmd/root.go:
```diff
@@ -92,15 +100,26 @@ func NewRootCmd(l zerolog.Logger) *cobra.Command {
+						logger.Debug().Err(err).Msg("sending command to telemetry")
+					}
+
+					telemetryWg.Done()
```
this probably should be called using `defer` `defer telemetryWg.Done()`

migmartri commented on app/cli/cmd/root.go:
```diff
@@ -92,15 +100,26 @@ func NewRootCmd(l zerolog.Logger) *cobra.Command {
+				logger.Debug().Msg("Telemetry enabled, to disable it use DO_NOT_TRACK=1")
+				// For telemetry reasons we parse the token to know the type of token is being used when executing the CLI
+				// Once we have the token type we can send it to the telemetry service by injecting it on the context
+				token, err := parseToken(apiToken)
```
just realized that this can probably be inside the `subroutine` too

migmartri commented on app/cli/internal/telemetry/telemetry.go:
```diff
@@ -49,28 +50,57 @@ func NewCommandTracker(client Client) *CommandTracker {
+}
+
+// DetermineUserID Determines the user ID using available information or generates a random one.
+func DetermineUserID(tags Tags) string {
```
`DetermineUserID` can be downcase if you move the tests to the same package

javirln replied:
I've added it into a separated packaged because if added to the same package it creates an import cycle between the mocks and the implementation:
```
FAIL	github.com/chainloop-dev/chainloop/app/cli/internal/telemetry [setup failed]
package github.com/chainloop-dev/chainloop/app/cli/internal/telemetry
	imports github.com/chainloop-dev/chainloop/app/cli/internal/telemetry/mocks
	imports github.com/chainloop-dev/chainloop/app/cli/internal/telemetry: import cycle not allowed in test
```

migmartri replied:
You are right, this is the problem of using the mocks generated by mockery :(

A workaround is to create another file, `telemerty_unit_test.go` or similar and have these unit tests. 

In general when I end up writing code in a specific way just so it works in the tests, I realize that I might be doing smth wrong and re-evaluate. But up to you. 

Thansk!

javirln replied:
>A workaround is to create another file, telemerty_unit_test.go or similar and have these unit tests.

But the problem would still be the same, the method will still be public right?

I mean for me, I wouldn't even create a test for this specific function since they are covered by the tests on the `Track` method that includes it.

jiparis commented on app/cli/cmd/root.go:
```diff
@@ -239,46 +268,125 @@ func loadControlplaneAuthToken(cmd *cobra.Command) (string, error) {
 // Each one of them have an associated audience claim that we use to identify the type of token. If the token is not
 // present, nor we cannot match it with one of the expected audience, return nil.
-func parseToken(token string) (*ParsedToken, error) {
+func parseToken(token string) (*parsedToken, error) {
```
Not for this PR, but most probably we can have a common interface. Something like 
```
type Token interface {
  func Parse(token string) (*parsedToken, error)
}
```
And then moving the "extracting" logic to their own implementations.

jiparis commented on app/cli/cmd/root.go:
```diff
@@ -239,46 +268,125 @@ func loadControlplaneAuthToken(cmd *cobra.Command) (string, error) {
+	return pToken, nil
+}
+
+var (
```
Not important: Do you think we can move the below code to its own file `telemetry.go` or similar just for readability?

javirln replied:
I would do it if we really support more telemetry and figure out the problem with the logger since right now it's coupled to it.
